syntax = "proto3";

package whisperx.asr;

// ASR (Automatic Speech Recognition) Service
// Compatible with NVIDIA Riva ASR protocol
service AsrService {
  // Streaming recognition for real-time audio
  rpc StreamingRecognize(stream StreamingRecognizeRequest) returns (stream StreamingRecognizeResponse);
  
  // Single request recognition for audio files
  rpc Recognize(RecognizeRequest) returns (RecognizeResponse);
}

// Configuration for recognition requests
message RecognitionConfig {
  // Audio encoding format
  enum AudioEncoding {
    ENCODING_UNSPECIFIED = 0;
    LINEAR_PCM = 1;
    FLAC = 2;
    MULAW = 3;
    ALAW = 4;
    OGG_OPUS = 5;
    MP3 = 6;
  }
  
  AudioEncoding encoding = 1;
  int32 sample_rate_hertz = 2;
  int32 audio_channel_count = 3;
  string language_code = 4;
  int32 max_alternatives = 5;
  
  // Model to use for recognition (e.g., "large-v3", "medium", "small")
  string model = 6;
  
  // Enable automatic punctuation
  bool enable_automatic_punctuation = 7;
  
  // Enable speaker diarization
  bool enable_speaker_diarization = 8;
  
  // Number of speakers (if known)
  int32 diarization_speaker_count = 9;
  
  // Min speakers for diarization range
  int32 min_speaker_count = 10;
  
  // Max speakers for diarization range
  int32 max_speaker_count = 11;
  
  // Enable word-level timestamps
  bool enable_word_time_offsets = 12;
  
  // Task type: transcribe or translate
  string task = 13;
  
  // Initial prompt for the model
  string initial_prompt = 14;
}

// Audio content in the request
message RecognitionAudio {
  oneof audio_source {
    bytes content = 1;  // Inline audio bytes
    string uri = 2;     // URI to audio file (not implemented yet)
  }
}

// Request for single-shot recognition
message RecognizeRequest {
  RecognitionConfig config = 1;
  RecognitionAudio audio = 2;
}

// Request for streaming recognition
message StreamingRecognizeRequest {
  oneof streaming_request {
    StreamingRecognitionConfig streaming_config = 1;
    bytes audio_content = 2;
  }
}

// Streaming-specific configuration
message StreamingRecognitionConfig {
  RecognitionConfig config = 1;
  bool interim_results = 2;  // Return partial results
}

// Word information with timing
message WordInfo {
  double start_time = 1;   // Start time in seconds
  double end_time = 2;     // End time in seconds
  string word = 3;         // The word itself
  float confidence = 4;    // Confidence score (0.0 - 1.0)
  int32 speaker_tag = 5;   // Speaker ID if diarization enabled
}

// Alternative transcription result
message SpeechRecognitionAlternative {
  string transcript = 1;
  float confidence = 2;
  repeated WordInfo words = 3;
}

// Recognition result for a segment
message SpeechRecognitionResult {
  repeated SpeechRecognitionAlternative alternatives = 1;
  int32 channel_tag = 2;
  string language_code = 3;
  double audio_processed = 4;  // Duration of audio processed in seconds
}

// Response for single-shot recognition
message RecognizeResponse {
  repeated SpeechRecognitionResult results = 1;
}

// Response for streaming recognition
message StreamingRecognizeResponse {
  repeated StreamingRecognitionResult results = 1;
}

// Streaming recognition result
message StreamingRecognitionResult {
  repeated SpeechRecognitionAlternative alternatives = 1;
  bool is_final = 2;
  float stability = 3;
  int32 channel_tag = 4;
  string language_code = 5;
}
